sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala

```diff
@@ -235,6 +200,56 @@ private[hive] case class HiveGenericUDF(
   }
 }
 
+class HiveGenericUDFHelper(
+    funcWrapper: HiveFunctionWrapper, children: Seq[Expression])
+  extends HiveInspectors
+  with Serializable
+  with Logging {
+
+  @transient
+  private[hive] val deterministic = isUDFDeterministic && children.forall(_.deterministic)
+
+  @transient
+  private[hive] val foldable =
+    isUDFDeterministic && returnInspector.isInstanceOf[ConstantObjectInspector]
+
+  @transient
+  private[hive] val dataType: DataType = inspectorToDataType(returnInspector)
+
+  @transient
+  private lazy val function = funcWrapper.createFunction[GenericUDF]()
+
+  @transient
+  private lazy val isUDFDeterministic = {
+    val udfType = function.getClass.getAnnotation(classOf[HiveUDFType])
+    udfType != null && udfType.deterministic() && !udfType.stateful()
+  }
+
+  @transient
+  private lazy val argumentInspectors = children.map(toInspector)
+
+  @transient
+  private lazy val deferredObjects: Array[DeferredObject] = argumentInspectors.zip(children).map {
```

Comment:
`lazy val` has perf overhead. I don't think this class needs any lazy val, but itself should be lazy val in `HiveGenericUDF`

Reply:
Ok

Reply:
when we reference the new evaluator, some properties in HiveGenericUDFEvaluator will have to use lazy, because it needs to be recreated on the Executor side.

Reply:
This makes me think that we should only keep execution-related things in this evaluator: `function` and `deferredObjects`. Can we expose `argumentInspectors` so that expression can define `dataType`, `foldable`, etc.?

Reply:
For expression can define dataType, foldable, deterministic, we will expose `argumentInspectors `, `returnInspector`.
Done